###### R script to visualize the most abundant microsatellite alleles from Illumina amplicon sequencing and the generation of a matrix based on allele
###### length and frequency distribution, generates a color-coded pdf with barplots of allele lengths and their relative frequency  
###### generated by S. Winter

## check commented lines number 54 for alpha now set to 0.62 and lines 84 and 103 for x-axis limits

#install.packages("reshape")
#install.packages("reshape2")
#library(reshape2)
#if(!require(reshape2)){install.packages("reshape") & install.packages("reshape2")}   #needed for command line call in bash script

#Passing over parameters from command line/ bash script
#cat(getwd(), "\n")
#setwd("/home/swinter/Igel_Microsat")
args <- commandArgs(TRUE)
path=(as.character(args[1]))        #path of directory of inputfiles with allele lengths and counts per marker and sample
mincount=as.numeric(args[2])        #minimum treshold of total read count to consider that marker in sample, absolute minimum 500 recommended, better higher
pathPDF=(as.character(args[3]))     #path to directory where to store pdf 
alpha=as.numeric(args[4])           #threshold for relative frequency per read length to be considered as homozygous
pathmatrix=(as.character(args[5]))     #path to directory where to store marker matrix 
pathcsvlist=(as.character(args[6]))   #path to directory where to store list of markers and samples with read length
a=(as.numeric(args[7]))             # lower x-axis limit for plots
b=(as.numeric(args[8]))          #upper x-axis limit for plots

print(a)
print(b)

inputfolder_name <- basename(path)

#define if the right input parameters have been provided - only necessary for mincount and alpha
#mincount=400
#if (is.numeric(mincount) != TRUE | (mincount>20) != TRUE) {stop("You have not provided a minimum number for read counts per marker and sample or
                                                             #       the number is too small. It should be larger than 20.")}
#alpha=0.4
if (is.numeric(alpha) != TRUE | (alpha<0.9) != TRUE | (alpha>0.5) != TRUE) {
  stop("You have not provided an adequate number for alpha, should not be smaller than 0.5 and not larger than 0.9.")}


#possibly enable facultative command line parameters for xlim upper and lower for plots
#a=as.numeric(args[5]) 
#b=as.numeric(args[6])
#a=100; b=950
#if (is.numeric(a) != TRUE | (a>500) != TRUE | (a<100) != TRUE) {
#  stop("You have not provided an adequate number for the lower x-axis limit above 100 bp read length and below 500 bp read length")}
#if (is.numeric(b) != TRUE | (b>900) != TRUE | (b<200) != TRUE) {
#  stop("You have not provided an adequate number for the upper x-axis limit above 200 bp read length and below 900 bp read length")}
#hard coded paths for developing the script:
##cat("path: ", path, "\n")
#path="~/Igel_Microsat/Outputdir17-09-12/MarkerStatistics/"
#pathPDF = "~/Igel_Microsat/Outputdir17-09-12/Markerplots/markerplots.pdf"
#pathmatrix = "~/Igel_Microsat/Outputdir17-09-12/Markerplots/markermatrix.csv"

############## SECTION for Running R directly - testing mode for data #############################################################
#Code needed to run Script from R directly - testing mode
#install.packages("reshape")
#install.packages("reshape2")
library(reshape2)
#hard coded paths for developing the script:
#cat("path: ", path, "\n")
#path="~/Igel_Microsat/Outputdir17-09-12/MarkerStatistics/"
#path="E:/Masterarbeit_Bioinformatik/InputData_RScript/MarkerStatistics_EasternAll/"     #take care that you separate hierarchy of folders by "/" 
#path="E:/Masterarbeit_Bioinformatik/InputData_RScript/TEST/"
#getwd()
#setwd("E:/Masterarbeit_Bioinformatik")
#pathPDF = "E:/Masterarbeit_Bioinformatik/Markerplots/markerplotsEasternAllBarySTUTTER.pdf"
#pathmatrix = "E:/Masterarbeit_Bioinformatik/Markerplots/markermatrixEasternAllBarySTUTTER.csv"
#mincount=20      #minimum treshold of total read count to consider that marker in sample, absolute minimum 500 recommended, better higher
#alpha=0.62         #threshold for relative frequency per read length to be considered as homozygous
#a=400            # lower x-axis limit for plots
#b=550             # upper x-axis limit for plots
##########################################################################################################################################################

#Generate pdf with bar charts per marker and sample, generate 6 plots per page 
start.time<- Sys.time() 
pdf(file=pathPDF,width=7,height=18)
par(mfrow=c(6,1))
op <- options(stringsAsFactors=F) 
dat<-data.frame()    #initialise empty dataframe for marker length information
i=0

#file="/home/swinter/Igel_Microsat/Outputdir17-09-07/MarkerStatistics//HH10_AAAG_IBK9_.statistics"     
#file="E:/Masterarbeit_Bioinformatik/InputData_RScript/MarkerStatistics_EasternAll/HH1_AC_2006432_.statistics"
#start.time<- Sys.time()    #check how long execution of for loop takes

i = 0
xlim_lower_smallest = 0
xlim_upper_largest = 0

if (a == 0 | b == 0){ #is.na(a) | is.na(b) if a or b is null value
  for (file in dir(path,full.names=TRUE)) {
    if (file.size(file) != 0 ){
      filein<-read.table(file, header=F,sep= " "); #filein
      x_min <- min(filein[,1])
      x_max <- max(filein[,1])
      if (i == 0) {
        xlim_lower_smallest <- x_min - 0.01 * x_min
        xlim_upper_largest <- x_max + 0.01 * x_max
      }
      else{
        xlim_lower <- x_min - 0.01 * x_min
        xlim_upper <- x_max + 0.01 * x_max
        if (xlim_lower < xlim_lower_smallest){
          xlim_lower_smallest <- xlim_lower
        }
        if (xlim_upper > xlim_upper_largest){
          xlim_upper_largest <- xlim_upper
        }
      }
      i <- i + 1
    }
  }
}else{
  xlim_lower_smallest = a
  xlim_upper_largest = b
}

sorted_files <- dir(path, full.names = TRUE)

custom_sort <- function(files) {
  info <- lapply(files, extract_info)
  sorted_indices <- order(sapply(info, function(x) x$samplename), sapply(info, function(x) x$marker)) #, sapply(info, function(x) x$repstring)
  return(files[sorted_indices])
}

extract_info <- function(filename) {
    parts <- unlist(strsplit(basename(filename), "_"))
    marker <- parts[1]
    repstring <- parts[2]
    samplename <- parts[3]
    return(list(marker = marker, repstring = repstring, samplename = samplename))
}

if (inputfolder_name == "MarkerStatisticsBoundaries"){
  sorted_files <- custom_sort(sorted_files)
}




for (file in sorted_files) {
  #print(paste ("filename is ", file))
  output<-basename(file.path(file,fsep=".statistics")); #output
  #print(paste ("basename is", output))
  stringsplit1<-sapply(strsplit(output, split = "_.statistics"),"[",1); #stringsplit1    #extract part of file name with sample name and marker name for matrix
  splits<-as.character(stringsplit1); #splits
  repstring<-sapply(strsplit(splits, split = "_"), "[", 2); #repstring
  samplename<-sapply(strsplit(splits, split = "_"), "[", 3); #samplename
  markern<-sapply(strsplit(splits, split = "_"), "[", 1); #markern   #Regexpression to extract markername
  #markern1<-substring(markern, 2); markern1
  markername<-paste(markern,repstring,sep="_"); #markername
  repsiz1<-nchar(repstring); #repsiz1                    #extract repeat size length from repeat motif
  repsiz<-as.integer(repsiz1); #repsiz
  if (file.size(file) == 0 ) {   
    #print(paste("empty files are: ", i, file,sep=" "))
    i=i+1
    plot(1,ylab="Count Frequency (%)", xlab="Allele Length", xlim=c(xlim_lower_smallest, xlim_upper_largest), ylim=c(0,105), type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is 0",sep=" "))
    mtext("empty file",side=3,cex=0.8)
    foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "empty")
    dat<-rbind(dat, foo)
    foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "empty")
    dat<-rbind(dat, foo)
  } else {   
  filein<-read.table(file,header=F,sep= " "); #filein
  maxcount<-sum(filein[,2])#; maxcount   #maxcount = total read count
  
  x_min <- min(filein[,1])
  x_max <- max(filein[,1])
  xlim_lower <- x_min - 0.01 * x_min
  xlim_upper <- x_max + 0.01 * x_max
  
  #Filter Implementation 
  #filter<-which(apply(filein, 1, function (x) all(x[2]/maxcount >= 0.05))) # with 1 reads in row vectors, all above a threshold of 5% maintained
  #fileinF<-filein[filter, ]
  #fileinF
  #maxcount2<-sum(fileinF[,2])  # new maxcount could be different as low quantity reads have been deleated
  
  maxPeak<-max(filein[,2]); maxPeak                  #highest read count of sample and marker
  #maxleng<-filein$V1[filein$V2==maxPeak]; maxleng
  maxleng<-tail(filein$V1,n=1); maxleng              #read length associated with highest read count
  #maxPeak2<-sort(filein$V2,decreasing=TRUE)[n-1]; maxPeak2
  #Here we should sustract difference between largest peak and the peak before it with the difference between them if maxleng - repsiz
  #a<-filein$V2[filein$V1==maxleng];a   #get the read count of largest lengt
  
  potstutt<-maxleng-repsiz
  potstutt
  potstutt2<-maxleng-2*repsiz; potstutt2
  filein1 <- filein
  stutt <- filein1[filein1$V1==potstutt,]
  if(repsiz == 2) {
    if(nrow(stutt) > 0) {
      if((stutt$V2 <= 0.85*maxPeak) & (stutt$V2 >= 0.35*maxPeak)) {
        stotter <- filein1[filein1$V1==potstutt,]$V2
        stutt2 <- filein1[filein1$V1==potstutt2,]
        if(nrow(stutt2) > 0) {
          if(stutt2$V2 <= (0.85*stotter) & (stutt$V2 >= 0.35*stotter)) {
            filein1[filein1$V1==potstutt2,]$V2 = filein1[filein1$V1==potstutt2,]$V2 * 0.65
            filein1[filein1$V1==potstutt,]$V2 = filein1[filein1$V1==potstutt,]$V2 * 0.6
          }
        }
      }
    }
  }

  filein1<-filein1[order(filein1$V2),]
  
  maxPeak2<-head(tail(filein1$V2,n=2),n=1); maxPeak2
  #maxleng2<-filein$V1[filein$V2==maxPeak2]; maxleng2             #Problem zwei ganz gleich haeufige allele
  maxleng2<-head(tail(filein1$V1,n=2),n=1); maxleng2
  n<-length(filein[,2]); n
  
  #Baryzentrische coordinates to decide if sample is heterozygous, homozygous or need manual control
  b1<-(maxPeak-maxPeak2)
  b2<-(maxPeak2*2)
  b3<-(1-(maxPeak+maxPeak2))

  duplicates <- FALSE
  if (ncol(filein) >= 4) { #&& (ncol(filein) - 2) %% 2 == 0
    duplicates <- TRUE
    #boundaries <- as.character(filein[,3]) boundaries
    # lower_boundary <- as.numeric(filein[1,3])  # Assuming second row is the lower boundary
    # upper_boundary <- as.numeric(filein[1,4])
    boundaries <- list()

    for (i in seq(3, ncol(filein), by = 2)) {
      # Extract the lower and upper boundary from consecutive columns
      if (i < ncol(filein)) {
        lower_boundary <- as.numeric(filein[1, i])
        upper_boundary <- as.numeric(filein[1, i + 1])

        # Append a list (acting as a tuple) of these boundaries to the boundaries list
        boundaries[[length(boundaries) + 1]] <- list(lower_boundary, upper_boundary)
      }
    }
  }
  print(duplicates)
  
  allelenames<-as.character(filein[,1]); allelenames
  countnames<-as.character(filein[,2]); countnames
  x_values <- filein[,1]
  y_values <- 100 * filein[,2] / maxcount
  
  lower_boundary = xlim_lower_smallest 
  upper_boundary = xlim_upper_largest
  # if (ncol(filein) >= 3) {
  #   #boundaries <- as.character(filein[,3]) boundaries
  #   lower_boundary <- as.numeric(filein[1,3])  # Assuming second row is the lower boundary
  #   upper_boundary <- as.numeric(filein[1,4])
  # }
	
  color_assignments <- ifelse(x_values < lower_boundary | x_values > upper_boundary, "black", "black")
  #print("Color assignments:")
  #print(table(color_assignments))
  
  #plot(x_values, y_values, type = "h", col = color_assignments, lwd = 5, xlim = c(xlim_lower_smallest, xlim_upper_largest), ylim = c(0, 105))

  if ((maxcount > mincount) & (n > 1)) {                               #call mincount variable from command line input here
          #first case one length highly abundant - homozygous because of baryzentric coordinates
              if (( b1 >= b2) & (b1 >= b3))  {
                  color_assignments[filein$V1 == maxleng] <- "blue"
                  plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Relative frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col = color_assignments) #col=ifelse(filein$V1==maxleng, "blue", "black")
                  text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames)
                  mtext(paste("Homozygous read count and allele is:", maxleng, ":", maxPeak, sep=" "),side=3, line=0, cex=0.8, col="blue")

                  if (duplicates) {
                      additional_message <- paste("Set Intervalls: ")
                      for (boundary_pair in boundaries) {
                          additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
                      }
                      mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
                  }

                  foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                  dat<-rbind(dat, foo)
                  foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng)
                  dat<-rbind(dat, foo)
                  
              } else if (repsiz != 0) {
          #heterozygous variants
                  #second case: heterozygous, most abundant and second most abundant larger than b1 and b3, difference in length according to repeat length, second highest allele longer than highest
                if ( (b2 >= b1) & (b2 >= b3) & ((abs(maxleng2-maxleng)) >= repsiz) & ((maxleng2-maxleng) > 0) & ((abs(maxleng2-maxleng)) %% repsiz == 0)) {    #second most abundant Allel länger als erstes 
                      color_assignments[filein$V1 == maxleng | filein$V1==maxleng2] <- "green"
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col = color_assignments) #col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"green", "black")
                      mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), side=3, line=0, col="dark green",cex=0.8) 
                      text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames, col="dark green")

                      if (duplicates) {
                          additional_message <- paste("Set Intervalls: ")
                          for (boundary_pair in boundaries) {
                              additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
                          }
                          mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
                      }

                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)

                  #third case - heterozygous, diff most abundant and second most in length more than 1 x repeat size (exclusion stutter), difference exactly integer x repeat size
              } else if ( (b2 >= b1) & (b2 >= b3) & (abs(maxleng2-maxleng) > repsiz) & (abs(maxleng2-maxleng) %% repsiz == 0) ) {
                      color_assignments[filein$V1==maxleng | filein$V1==maxleng2] <- "red"
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col = color_assignments) #col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"red", "black")
                      mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, line=0, side=3, col="black") 
                      text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames, col="black")

                      if (duplicates) {
                          additional_message <- paste("Set Intervalls: ")
                          for (boundary_pair in boundaries) {
                              additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
                          }
                          mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
                      }

                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
          
                  # fourth case - point mutations that are minimum size of 60% max peak but don't match repeat size, HERE change due to FILTER
              } else if ( (b2 >= b1) & (b2 >= b3) & ((maxPeak2/maxcount) > (0.6*maxPeak/maxcount)) & ((abs(maxleng2-maxleng)) %% repsiz != 0) )   {  #second peak needs to minimum read count of 0.6xhighest peak   
                      color_assignments[filein$V1==maxleng | filein$V1==maxleng2] <- "red"                                             
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"red", "black"))
                      mtext(paste("Possibly heterozygous point mutation, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, line=0, side=3,col="green")
                      text(filein[,1]-0.5, 100*filein[,2]/maxcount+5 ,allelenames, col="black")

                      if (duplicates) {
                          additional_message <- paste("Set Intervalls: ")
                          for (boundary_pair in boundaries) {
                              additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
                          }
                          mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
                      }  

                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
                      
                 # fifth case - only one repeat size difference but second largest peak is minimum size of 0.75 % of the largest one,    
              } else if ( (b2 >= b1) & (b2 >= b3) & ((maxPeak2/maxcount) > (0.75*maxPeak/maxcount)) & (abs(maxleng2-maxleng) == repsiz)  )   { 
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"red", "black"))
                      mtext(paste("Heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, line=0, side=3, col="black") 
                      text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames, col="black")

                      if (duplicates) {
                          additional_message <- paste("Set Intervalls: ")
                          for (boundary_pair in boundaries) {
                              additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
                          }
                          mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
                      }

                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
              } else {
                # Manual check for repsiz != 0
                plot(filein[,1], 100 * filein[,2] / maxcount, xlab = "Allele Length", ylab = "Count Frequency (%)", type = "h", main = paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub = paste("Total read count is", maxcount, sep = " "), lwd = 5, xlim = c(xlim_lower_smallest, xlim_upper_largest), ylim = c(0, 105), col = ifelse(filein$V1 == maxleng | filein$V1 == maxleng2, "red", "black"))
                mtext(paste("Needs manual check, read counts of highest peaks are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep = " "), side = 3, line = 0, col = "black", cex = 0.8)
                text(filein[,1] - 0.01, 100 * filein[,2] / maxcount + 5, allelenames, col = "black")
                
                if (duplicates) {
                  additional_message <- paste("Set Intervalls: ")
                  for (boundary_pair in boundaries) {
                    additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep = "")
                  }
                  mtext(additional_message, side = 3, line = 3, cex = 0.8, col = "black")
                }
                
                foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep = "_"), maxleng = paste(maxleng, "man check", sep = "_"))
                dat <- rbind(dat, foo)
                foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep = "_"), maxleng = paste(maxleng2, "man check", sep = "_"))
                dat <- rbind(dat, foo)
                
              }
              } else if (repsiz == 0) {
                if ((b2 >= b1) & (b2 >= b3) & ((maxPeak2 / maxcount) > (0.6 * maxPeak / maxcount))) {
                  color_assignments[filein$V1 == maxleng | filein$V1 == maxleng2] <- "red"
                  plot(filein[,1], 100 * filein[,2] / maxcount, xlab = "Allele Length", ylab = "Count Frequency (%)", type = "h", main = paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub = paste("Total read count is", maxcount, sep = " "), lwd = 5, xlim = c(xlim_lower_smallest, xlim_upper_largest), ylim = c(0, 105), col = ifelse(filein$V1 == maxleng | filein$V1 == maxleng2, "red", "black"))
                  mtext(paste("Possibly heterozygous point mutation, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep = " "), cex = 0.8, line = 0, side = 3, col = "green")
                  text(filein[,1] - 0.5, 100 * filein[,2] / maxcount + 5, allelenames, col = "black")
                  
                  if (duplicates) {
                    additional_message <- paste("Set Intervalls: ")
                    for (boundary_pair in boundaries) {
                      additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep = "")
                    }
                    mtext(additional_message, side = 3, line = 3, cex = 0.8, col = "black")
                  }
                  
                  foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep = "_"), maxleng = maxleng)
                  dat <- rbind(dat, foo)
                  foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep = "_"), maxleng = maxleng2)
                  dat <- rbind(dat, foo)
                } else {
                  # Manual check for repsiz == 0
                  plot(filein[,1], 100 * filein[,2] / maxcount, xlab = "Allele Length", ylab = "Count Frequency (%)", type = "h", main = paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub = paste("Total read count is", maxcount, sep = " "), lwd = 5, xlim = c(xlim_lower_smallest, xlim_upper_largest), ylim = c(0, 105), col = ifelse(filein$V1 == maxleng | filein$V1 == maxleng2, "red", "black"))
                  mtext(paste("Needs manual check, read counts of highest peaks are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep = " "), side = 3, line = 0, col = "black", cex = 0.8)
                  text(filein[,1] - 0.01, 100 * filein[,2] / maxcount + 5, allelenames, col = "black")
                  
                  if (duplicates) {
                    additional_message <- paste("Set Intervalls: ")
                    for (boundary_pair in boundaries) {
                      additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep = "")
                    }
                    mtext(additional_message, side = 3, line = 3, cex = 0.8, col = "black")
                  }
                  
                  foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep = "_"), maxleng = paste(maxleng, "man check", sep = "_"))
                  dat <- rbind(dat, foo)
                  foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep = "_"), maxleng = paste(maxleng2, "man check", sep = "_"))
                  dat <- rbind(dat, foo)
                }
              }
  } else if ((maxcount > mincount) & (n = 1)) { 
        plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col=ifelse(filein$V1==maxleng,"blue", "black"))
        mtext(paste("Homozygous, only one read length present:", maxleng, ":", maxPeak, sep=" "),side=3, line=0, col="black", cex=0.8)
        text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames, col="black")

        if (duplicates) {
            additional_message <- paste("Set Intervalls: ")
            for (boundary_pair in boundaries) {
                additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
            }
            mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
        }

        foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
        dat<-rbind(dat, foo)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng)
        dat<-rbind(dat, foo)
  } else if (maxcount == 0) {
        plot(1,ylab="empty file", xlab="", xlim=c(a, b), ylim=c(0,105), type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is 0",sep=" "))
        mtext("empty file",side=3,cex=0.8)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "empty")
        dat<-rbind(dat, foo)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "empty")
        dat<-rbind(dat, foo)
  } else {
     plot(filein[,1],100*filein[,2]/maxcount,xlab="Allele Length",ylab="Count Frequency (%)",type="h", main=paste("Sample:", samplename, ", Marker:", markername, ", Filename: ", markername, "_", samplename, "_.txt"), sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(xlim_lower_smallest,xlim_upper_largest),ylim=c(0,105), col="black")
     mtext(paste("Too little read count:", maxleng, ":", maxPeak, sep=" "),side=3, line=0, cex=0.8, col="black")
     text(filein[,1]-0.01, 100*filein[,2]/maxcount+5 ,allelenames, col="black")

    if (duplicates) {
        additional_message <- paste("Set Intervalls: ")
        for (boundary_pair in boundaries) {
            additional_message <- paste(additional_message, sprintf("(%d, %d)", boundary_pair[[1]], boundary_pair[[2]]), sep="")
        }
        mtext(additional_message, side=3, line=3, cex=0.8, col="black")  # line=1 places this below the first
    }

     foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "too little reads")
     dat<-rbind(dat, foo)
     foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "too little reads")
     dat<-rbind(dat, foo)
  }
  }
}
dev.off()
end.time<-Sys.time()
time.taken<-end.time - start.time
time.taken

#if (inputfolder_name != "MarkerStatisticsBoundaries"){}

dat
dat<- melt(dat,id.vars= c("samplename","marker"))
res<-dcast(dat,samplename~marker,function(x){x[1]})
res
#write.csv(res,file=pathmatrix,row.names=FALSE)
write.table(res,file=pathmatrix,row.names=FALSE,sep =";")


#options(op)  #wegen Fehlermeldungen nötig: 0: In `[<-.factor`(`*tmp*`, ri, value = 437L) : ungültiges Faktorniveau, NA erzeugt
